### PR TITLE
Change GRPC message limit SQL error type

### PR DESCRIPTION
**Pull request by setassociative**
## Summary

There are a collection of errors bundled under the generic GRPC ResourceExhausted error. Several of them are not necessarily related to resource exhausted but to limits that are being exceeded in a way that backing off and waiting for reduced load will not resolve.

This change introduces a processing step in the GRPC -> mysql error conversion that rewrites those errors from the current `1203 - ERTooManyUserConnections` to a (somewhat) more appropriate `1153 - ERNetPacketTooLarge`. Two other errors are also recategorized in the vtgate executor to be treated as the 1153 instead of 1203.

The three errors being changed:

- When a GRPC message coming back to the vtgate exceeds the configured size limit (configured via `grpc_max_message_size`)
- If a write is attempting to make a change with a payload that is larger than the configured allowable size
- if a scatter query is attempting to collect process too many rows in memory within the vtgate to construct a response

Before this change each of these would the SQL error `ERTooManyUserConnections` when using the MySQL interface. This is a fairly unfortunate conflation of that error and doesn't make sense to be consumed as "too many users" when the actual error is various flavors of "too much data."

Further in a situation where you want to use error codes to apply back pressure to callers having these two classes of error combined means that it's not a great signal.

We discussed briefly whether we should _also_ change the GRPC error so that ResourceExhausted isn't overloaded but it seems that those errors need to match GRPC error set ([ref](https://github.com/vitessio/vitess/blob/f886cc2681844f3bec3dd3e9b248e22b24c30a92/proto/vtrpc.proto#L53-L55)). Additionally it would be a much larger change because ResourceExhausted is used for some session management. If we decide later to revisit this decision that will be fine as the GRPC and SQL errors are independent (also, even if it does feel like what we'd want to do I would prefer it to be a separate change from this from a pragmatic standpoint).

### Misc

1️⃣  This change exceeds our initial plan which covered only the GRPC message limit case and I'm open for pushback on the other two but I do think that all these cases are better suited as 1153 than 1203.

2️⃣ The error message checks are _extremely factored_ and I'm happy to pull them out of `vterrors` for another home or deal with that differently if folks have opinions. I didn't know what was the most reasonable location since they needed to be shared between sql_errors and the vtgate

## Testing

- Added unit tests for the method that processes the resource exhausted errors
- Built and deployed a modified vtgate and observed expected (new) error code when running with an artificially low grpc message limit (3k). Note `ERROR 1153` in the following:

```
mysql> select id, data from test_table limit 100;
ERROR 1153 (HY000): vtgate: http://vtgate-dev/: target: keyspace.-.master, used tablet: <tablet_uid> (<tablet-host>): vttablet: rpc error: code = ResourceExhausted desc = grpc: received message larger than max (12579 vs. 3072)
```

- Observed row-count limit related errors returning 1153 mysql error:
```
mysql> select id, data from test_table limit 10000;
ERROR 1153 (HY000): vtgate: http://vtgate-dev/: in-memory row count exceeded allowed limit of 100
```

- Explicitly: did not test the payload-too-large case and relying on unit tests 🤠 

## Reference

We talked about this in the [Vitess slack](https://vitess.slack.com/archives/CMDJ2KFEZ/p1588031472083800) and decided on `ERNetPacketTooLarge` as the new error code there.

## Release Note Required

A MySQL error code 1153 is now being returned in the following cases:
- When a GRPC message coming back to the vtgate exceeds the configured size limit
- If a write is attempting to make a change with a payload that is larger than the configured allowable size
- if a scatter query is attempting to process too many rows in memory within the vtgate constructing the result set

Previously these would be returned as MySQL 1203 if received over the MySQL connector or `Code_RESOURCE_EXHAUSTED` over GRPC.

### CODE REVIEW

File: go/mysql/sql_error.go [hunk 1]
```diff
@@ -161,3 +161,12 @@ func NewSQLErrorFromError(err error) error {
 	}
 	return serr
 }
+
+var isGRPCOverflowRE = regexp.MustCompile(`.*grpc: received message larger than max \(\d+ vs. \d+\)`)
+
+func demuxResourceExhaustedErrors(msg string) int {
```
Comment (setassociative) [inline go/mysql/sql_error.go:167]: TODO: In addition to processing returned errors we should probably also follow this pattern when we construct payload size errors - https://github.com/vitessio/vitess/blob/master/go/vt/vtgate/executor.go#L1131

Comment (setassociative) [reply go/mysql/sql_error.go:167]: Ah, actually I believe because these (seem to be) expecting a `vtrpc.Code` style error instead of SQL code we can't actually do this here.

Comment (setassociative) [reply go/mysql/sql_error.go:167]: I'm just having a conversation with myself at this point bc I like the github format for exploring the PR.

But I believe we should expand the `NewSQLErrorFromError` to handle these cases instead. Since (presumably) everything that gets adapted comes back through there it should be a valid chokepoint for converting these.

File: go/mysql/sql_error_test.go [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+		testCase{"grpc: received message larger than max (99282+ vs. 1234): trailer", ERTooManyUserConnections},
+		testCase{"grpc: received message larger than max (1234 vs. 1234)", ERNetPacketTooLarge},
+		testCase{"header: grpc: received message larger than max (1234 vs. 1234)", ERNetPacketTooLarge},
+		testCase{"query payload size above threshold", ERNetPacketTooLarge},
+		testCase{"in-memory row count exceeded allowed limit of 13", ERNetPacketTooLarge},
+	}
+
+	for _, c := range cases {
+		got := demuxResourceExhaustedErrors(c.msg)
+		if got != c.want {
```
Comment (systay) [inline go/mysql/sql_error_test.go:40]: nit: better to use `assert.Equal` for these comparisons

Comment (setassociative) [reply go/mysql/sql_error_test.go]: ✅

File: go/vt/vtgate/executor.go [hunk 3]
```diff
@@ -1128,7 +1128,7 @@ func (e *Executor) getPlan(vcursor *vcursorImpl, sql string, comments sqlparser.
 	statement := stmt
 	bindVarNeeds := sqlparser.BindVarNeeds{}
 	if !sqlparser.IgnoreMaxPayloadSizeDirective(statement) && !isValidPayloadSize(query) {
-		return nil, vterrors.New(vtrpcpb.Code_RESOURCE_EXHAUSTED, "query payload size above threshold")
+		return nil, vterrors.New(vtrpcpb.Code_RESOURCE_EXHAUSTED, vterrors.ResourceExhaustedQueryPayloadThresholdErrMsg)
```
Comment (harshit-gangal) [inline go/vt/vtgate/executor.go:1131]: Instead of this you can create sqlError here and then no changed would be needed in sql_error.go. It will return the sqlError as-is.

```
Example:
mysql.NewSQLError(mysql.ERBadDb, "42000", "Unknown database '%s'", keyspace)
```

File: go/vt/vtgate/scatter_conn.go [hunk 4]
```diff
@@ -235,7 +235,7 @@ func (stc *ScatterConn) ExecuteMultiShard(
 	)
 
 	if !ignoreMaxMemoryRows && len(qr.Rows) > *maxMemoryRows {
-		return nil, []error{vterrors.Errorf(vtrpcpb.Code_RESOURCE_EXHAUSTED, "in-memory row count exceeded allowed limit of %d", *maxMemoryRows)}
+		return nil, []error{vterrors.Errorf(vtrpcpb.Code_RESOURCE_EXHAUSTED, vterrors.ResourceExhaustedInMemoryLimitExceededFmt, *maxMemoryRows)}
```
Comment (harshit-gangal) [inline go/vt/vtgate/scatter_conn.go:238]: Same as above.

File: go/vt/vtgate/executor.go (hunk 3 above)
Comment (setassociative) [reply go/vt/vtgate/executor.go:1131]: Ah, nice -- it wasn't clear that we wanted to change this for all callers or how the executor would deal with getting a sql error back.

Comment (setassociative) [reply go/vt/vtgate/executor.go]: ✅

File: go/vt/vtgate/scatter_conn.go (hunk 4 above)
Comment (setassociative) [reply go/vt/vtgate/scatter_conn.go]: ✅

**Comment by setassociative** [comment]
@harshit-gangal I applied suggested error code changes; I believe the unit test that is failing is unrelated since, afaict, it's erroring out with relation to failing to connect to a mysql instance. Possible due to bad auth?

**Comment by deepthi** [comment]
@setassociative this could break applications that depend on the exact error code? Can you summarize the net effect in a `Release Note Required` section so that we can put it into the release notes?

File: go/mysql/sql_error.go [hunk 5]
```diff
@@ -120,7 +120,7 @@ func NewSQLErrorFromError(err error) error {
 		case vtrpcpb.Code_UNAUTHENTICATED:
 			num = ERAccessDeniedError
 		case vtrpcpb.Code_RESOURCE_EXHAUSTED:
-			num = ERTooManyUserConnections
+			num = demuxResourceExhaustedErrors(err.Error())
```
Comment (harshit-gangal) [inline go/mysql/sql_error.go:123]: is this still required after changing to through mysql error directly?

File: go/mysql/sql_error_test.go [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+	type testCase struct {
+		msg  string
+		want int
+	}
+
+	cases := []testCase{
+		testCase{"misc", ERTooManyUserConnections},
+		testCase{"grpc: received message larger than max (99282 vs. 1234): trailer", ERTooManyUserConnections},
+		testCase{"grpc: received message larger than max (1234 vs. 1234)", ERNetPacketTooLarge},
+		testCase{"header: grpc: received message larger than max (1234 vs. 1234)", ERNetPacketTooLarge},
```
Comment (harshit-gangal) [inline go/mysql/sql_error_test.go:35]: all of them should say ERNetPacketTooLarge. you should change the pattern matcher. I do not see a reason why the first test case should not say ERNetPacketTooLarge

File: go/mysql/sql_error.go (hunk 5 above)
Comment (setassociative) [reply go/mysql/sql_error.go:123]: Yes -- if we directly map all ResourceExhausted into ERNetPacketTooLarge then we end up in the same position where we are grouping errors that don't necessarily make sense together.

Specifically when we overflow a connection pool (`vttablet: rpc error: code = ResourceExhausted desc = pool ConnPool waiter count exceeded`) and the like we want to distinguish that as something that can be solved via backoff instead of queries that are inherently less retryable because they are likely to generate "bad" data.

File: go/mysql/sql_error_test.go (hunk 6 above)
Comment (setassociative) [reply go/mysql/sql_error_test.go:35]: Sure thing

**Review by harshit-gangal — APPROVED** [review]
